ebpf/python/pyperf.go

```diff
@@ -0,0 +1,264 @@
+	s.pidCache.Add(pid, data)
+	return nil
+}
+
+func (s *Perf) loop() {
+	defer s.rd.Close()
+
+	for {
+		record, err := s.rd.Read()
+		if err != nil {
+			if errors.Is(err, perf.ErrClosed) {
+				return
+			}
+			_ = level.Error(s.logger).Log("msg", "[pyperf] reading from perf event reader", "err", err)
+			continue
+		}
+
+		if record.LostSamples != 0 {
+			s.metrics.LostSamples.Add(float64(record.LostSamples))
+			_ = level.Debug(s.logger).Log("msg", "[pyperf] perf event ring buffer full, dropped samples", "n", record.LostSamples)
+		}
+
+		if record.RawSample != nil {
+			event, err := ReadPyEvent(record.RawSample)
+			if err != nil {
+				_ = level.Error(s.logger).Log("msg", "[pyperf] parsing perf event record", "err", err)
+				continue
+			}
+			s.eventsLock.Lock()
+			s.events = append(s.events, event)
+			s.eventsLock.Unlock()
+		}
+	}
+
+}
+
+func (s *Perf) Close() {
+	_ = s.rd.Close()
+	s.wg.Wait()
+}
+
+func (s *Perf) ResetEvents() []*PerfPyEvent {
+	s.eventsLock.Lock()
+	defer s.eventsLock.Unlock()
+	if len(s.events) == 0 {
+		return nil
+	}
+	eventsCopy := make([]*PerfPyEvent, len(s.events))
+	copy(eventsCopy, s.events)
+	for i := range s.events {
+		s.events[i] = nil
+	}
+	s.events = s.events[:0]
+
+	return eventsCopy
+}
+
+func (s *Perf) GetLazySymbols() LazySymbols {
+	return LazySymbols{
+		symbols: s.prevSymbols,
+		fresh:   false,
+		perf:    s,
+	}
+}
+
+func (s *Perf) GetSymbols(svcReason string) (map[uint32]*PerfPySymbol, error) {
+	s.metrics.SymbolLookup.WithLabelValues(svcReason).Inc()
+	var (
+		m       = s.symbolsHashMp
+		mapSize = m.MaxEntries()
+		nextKey = PerfPySymbol{}
+	)
+	keys := make([]PerfPySymbol, mapSize)
+	values := make([]uint32, mapSize)
+	res := make(map[uint32]*PerfPySymbol)
+	opts := &ebpf.BatchOptions{}
+	n, err := m.BatchLookup(nil, &nextKey, keys, values, opts)
+	if n > 0 {
+		level.Debug(s.logger).Log(
+			"msg", "GetSymbols BatchLookup",
+			"count", n,
+		)
+		res := make(map[uint32]*PerfPySymbol, n)
+		for i := 0; i < n; i++ {
+			k := values[i]
+			res[k] = &keys[i]
+		}
+		s.prevSymbols = res
+		return res, nil
+	}
+	if errors.Is(err, ebpf.ErrKeyNotExist) {
+		return nil, nil
+	}
+	// batch not supported
+
+	// try iterating if batch failed
+	it := m.Iterate()
+
+	v := uint32(0)
+	for {
+		k := new(PerfPySymbol)
+		ok := it.Next(k, &v)
+		if !ok {
+			err := it.Err()
+			if err != nil {
+				err = fmt.Errorf("map %s iteration : %w", m.String(), err)
+				return nil, err
+			}
+			break
+		}
+		res[v] = k
+	}
+	level.Debug(s.logger).Log(
+		"msg", "GetSymbols iter",
+		"count", len(res),
+	)
+	s.prevSymbols = res
+	return res, nil
+}
+
+func (s *Perf) RemoveDeadPID(pid uint32) {
+	s.pidCache.Remove(pid)
+	err := s.pidDataHashMap.Delete(pid)
+	if err != nil && !errors.Is(err, ebpf.ErrKeyNotExist) {
+		_ = level.Error(s.logger).Log("msg", "[pyperf] deleting pid data hash map", "err", err)
+	}
+}
+
+func ReadPyEvent(raw []byte) (*PerfPyEvent, error) {
```

Comment:
I would use the pattern as json here

``
 ReadPyEvent(data []byte, v *PerfPyEvent) error
```

This will be easier for the caller to reuse the object.

Reply:
applied the function prototype change, not reusing the event just yet